src/ws/cockpitcertificate.c

```diff
@@ -380,21 +380,39 @@ cockpit_certificate_load (const gchar *cert_path,
                           GError **error)
 {
   int r;
-  g_autofree gchar *certs = NULL;
-  g_autofree gchar *key = NULL;
-  g_autofree gchar *combined = NULL;
+  g_autofree gchar *key_path = cockpit_certificate_key_path (cert_path);
   GTlsCertificate *cert;
 
-  r = cockpit_certificate_parse (cert_path, &certs, &key);
-  if (r < 0)
+  g_assert (error);
+
+  /* check if we have a separate .key file */
+  cert = g_tls_certificate_new_from_files (cert_path, key_path, error);
+  if (cert)
     {
-      g_set_error (error, G_IO_ERROR, g_io_error_from_errno (-r), "Failed to load %s: %s", cert_path, g_strerror (-r));
-      return NULL;
+      g_debug ("loaded separate cert %s and key %s", cert_path, key_path);
+    }
+  else if (g_error_matches (*error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
```

Comment:
The error handling in the glib case is so much nicer...

You shouldn't be using `*error` in this way . It's possible that someone calls `cockpit_certificate_load()` with `error` set to `NULL`.  You should make your own `my_error` or something for this sort of thing and propagate it at the end using `g_propagate_error()`

Reply:
That's why I added the `g_assert (error)`, but indeed with propagation that's nicer. Thanks!